nextcloud/server
# [stable27] fix(files): race condition on web files view change

* Resolves: #39565

## Summary

A small regression from https://github.com/nextcloud/server/pull/35772/

This issue does not exist on Nextcloud 28 due to full files to view migration.

### How to reproduce

1. Share "Folder A" with some "User"
2. Log in as "User"
3. Open Files
4. Open "Shared with you"
5. Click on "Folder A"

- **Expected behavior:**
  - it opens the sahred folder
  - only "/FolderA" files are requested in "Network"
- **Actual behavior:**
  - root files are shown for a second or as the result
  - "/" and "/FolderA" twice are requested

Try several times to reproduce. Then more files in root then more chances to reproduce.

https://github.com/nextcloud/server/assets/25978914/a846313d-a666-41c7-a7c1-beeefc3483c0

### Root of the issue

There is a race condition.

Opening a shared folder makes 2 actions:

https://github.com/nextcloud/server/blob/6bfbb475ff376bc5f9bafeb7753b5a5050016be6/apps/files_sharing/js/app.js#L260-L263

Changing view from "sharingin" to "files" emits `files:legacy-navigation:changed`

https://github.com/nextcloud/server/blob/6bfbb475ff376bc5f9bafeb7753b5a5050016be6/apps/files/js/app.js#L223-L226

Then the Files view is open on root path `/` which causes **fetching the root files list**.

https://github.com/nextcloud/server/blob/6bfbb475ff376bc5f9bafeb7753b5a5050016be6/apps/files/src/views/Navigation.vue#L233-L242

See `this.show`:

https://github.com/nextcloud/server/blob/2aa15a1afe4ce99c60e8a721d4895e8339c0275e/apps/files/src/views/Navigation.vue#L214-L219

And the handler:

https://github.com/nextcloud/server/blob/2aa15a1afe4ce99c60e8a721d4895e8339c0275e/apps/files/js/filelist.js#L779-L788

Here is were the first request is made.

Then the initial `changeDirectory` is called. Which also results in changing directory (to the correct one) and **fetching the actual shared directory files list**.

🏎️ Here is a race. When the root directory is large, its response may come later so root files are shown instead of the shared folder.

---

Changing view also triggers a `shown` event that sets the current directory to `/` and reloads content. It causes duplicated file requests.

https://github.com/nextcloud/server/blob/2aa15a1afe4ce99c60e8a721d4895e8339c0275e/apps/files/js/filelist.js#L765-L774

Before Vue migration, there was a check for `this.shown` that avoided this behavior.

https://github.com/nextcloud/server/blob/f720d6e73065bc8ad84317981f301f085a600667/apps/files/js/filelist.js#L761-L773

If I'm not wrong, with migrating navigation to Vue, we can just remove this part. cc @skjnldsv 

## TODO

- [x] Add `silent` mode support for legacy navigation changed. Actually, the same that was there before migration to Vue.
- [x] Remove resetting and reloading on view show

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
